corehq/apps/data_dictionary/views.py

```diff
@@ -103,20 +105,48 @@ def data_dictionary_json(request, domain, case_type_name=None):
         p = {
             "name": case_type.name,
             "fhir_resource_type": fhir_resource_type_name_by_case_type.get(case_type),
+            "groups": [],
             "properties": [],
         }
-        for prop in case_type.properties.all():
-            p['properties'].append({
-                "description": prop.description,
-                "label": prop.label,
-                "index": prop.index,
-                "fhir_resource_prop_path": fhir_resource_prop_by_case_prop.get(prop),
-                "name": prop.name,
-                "data_type": prop.data_type,
-                "group": prop.group_name,
-                "deprecated": prop.deprecated,
-                "allowed_values": {av.allowed_value: av.description for av in prop.allowed_values.all()},
-            })
+        for group in case_type.groups.all():
+            group_ret = {
+                "id": group.id,
+                "name": group.name,
+                "description": group.description,
+                "deprecated": group.deprecated,
+                "properties": []
+            }
+
+            group_properties = sorted(
+                [prop for prop in case_type.properties.all() if prop.group_name == group.name],
```

Comment:
we can't do `group.properties` directly?

Reply:
is there a way to fetch properties with grouping by group, when fetching records itself and so avoid doing it later?

Reply:
> that caused the properties to go out of order

And then just ordering by index in code, isnt that still better than fetching all properties and then filtering for ones for the group?

> For fetching properties without groups and with groups all in one query. We might need to add some nested Prefetch statements.

And that is not good because?
Just confirming that I meant fetching all properties of a case type, and have them grouped by the group, in the DB query itself.

Reply:
>And then just ordering by index in code, isnt that still better than fetching all properties and then filtering for ones for the group?

All the properties for the case_type are already prefetched through the Prefetch queries so on `case_type.properties.all()` it is not refetched.

>And that is not good because?
Just confirming that I meant fetching all properties of a case type, and have them grouped by the group, in the DB query itself.

The code that i have posted above fetches properties without group for a case_type and then fetches properties for groups in a nested query. This way on using `group.properties.all()` the properties remain in order, although this way the no of nested queries increases as it would also need to refetch the allowed values. 

Going with the grouping approach seems way better as you suggested. I will make the changes for that.